src/annotator/util/buckets.js

```diff
@@ -0,0 +1,248 @@
+ * @return {BucketInfo}
+ */
+export function buildBuckets(points) {
+  const buckets = /** @type {Array<Anchor[]>} */ (new Array());
+  const bucketPositions = /** @type {number[]} */ (new Array());
+
+  // Anchors that are part of the currently-being-built bucket, and a correspon-
+  // ding count of unclosed top edges seen for that anchor
+  const current = /** @type {{anchors: Anchor[], counts: number[] }} */ ({
+    anchors: new Array(),
+    counts: new Array(),
+  });
+
+  points.forEach((point, index) => {
+    const [position, delta, anchor] = point;
+    // Does this point represent the top or the bottom of an anchor's highlight
+    // box?
+    const positionType = delta > 0 ? 'start' : 'end';
+
+    // See if this point's anchor is already in our working set of open anchors
+    const anchorIndex = current.anchors.indexOf(anchor);
+
+    if (positionType === 'start') {
+      if (anchorIndex === -1) {
+        // Add an entry for this anchor to our current set of "open" anchors
+        current.anchors.unshift(anchor);
+        current.counts.unshift(1);
+      } else {
+        // Increment the number of times we've seen a start/top edge for this
+        // anchor
+        current.counts[anchorIndex]++;
+      }
+    } else {
+      // positionType = 'end'
+      // This is the bottom/end of an anchor that we should have already seen
+      // a top edge for. Decrement the count, representing that we've found an
+      // end point to balance a previously-seen start point
+      current.counts[anchorIndex]--;
+      if (current.counts[anchorIndex] === 0) {
+        // All start points for this anchor have been balanced by end point(s)
+        // So we can remove this anchor from our collection of open anchors
+        current.anchors.splice(anchorIndex, 1);
+        current.counts.splice(anchorIndex, 1);
+      }
+    }
+
+    // For each point, we'll either:
+    // * create a new bucket: Add a new bucket (w/corresponding bucket position)
+    //   and add the working anchors to the new bucket. This, of course, has
+    //   the effect of making the buckets collection larger. OR:
+    // * merge buckets: In most cases, merge the anchors from the last bucket
+    //   into the penultimate (previous) bucket and remove the last bucket (and
+    //   its corresponding `bucketPosition` entry). Also add the working anchors
+    //   to the previous bucket. Note that this decreases the size of the
+    //   buckets collection.
+    // The ultimate set of buckets is defined by the pattern of creating and
+    // merging/removing buckets as we iterate over points.
+    const isFirstOrLastPoint =
+      bucketPositions.length === 0 || index === points.length - 1;
+
+    const isLargeGap =
+      bucketPositions.length &&
+      position - bucketPositions[bucketPositions.length - 1] > BUCKET_GAP_SIZE;
+
+    if (current.anchors.length === 0 || isFirstOrLastPoint || isLargeGap) {
+      // Create a new bucket, because:
```

Comment:
I suppose this comment is a bit awkward because the latter two are self-documenting by the reference name...